File: app/src/main/java/com/example/librewards/TimerFragment.java
package com.example.librewards;

import android.app.Dialog;
import android.content.Context;
import android.content.SharedPreferences;
import android.database.Cursor;
import android.graphics.Color;
import android.graphics.drawable.ColorDrawable;
import android.icu.util.TimeUnit;
import android.os.Bundle;
import android.os.SystemClock;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.Chronometer;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.fragment.app.Fragment;

import java.sql.Time;
import java.time.Duration;
import java.util.ArrayList;
import java.util.List;

public class TimerFragment extends Fragment {
    Dialog popup;
    Dialog namePopup;
    Chronometer stopwatch;
    DatabaseHelper myDb;

    public static final String TAG = TimerFragment.class.getSimpleName();

    private ListFromFile listFromFile;
    public List<String> currStartCodes = new ArrayList<>();
    public List<String> originalStartCodes = new ArrayList<>();
    public List<String> currStopCodes = new ArrayList<>();
    public List<String> originalStopCodes = new ArrayList<>();

    private EditText editText;
    private String textToEdit;
    private Button startButton;
    private Button stopButton;
    private TextView points;
    private TextView name;
    TimerListener listener;

    public interface TimerListener {
        void onPointsTimerSent(int points);
    }
    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment

        View v = inflater.inflate(R.layout.fragment_timer, container, false);
        stopwatch = v.findViewById(R.id.stopwatch);
        editText = v.findViewById(R.id.startText);
        startButton = v.findViewById(R.id.startButton);
        stopButton = v.findViewById(R.id.stopButton);
        myDb = new DatabaseHelper(getActivity().getApplicationContext());
        points = v.findViewById(R.id.points);
        points.setText(String.valueOf(myDb.getPoints()));
        name = v.findViewById(R.id.nameTimer);
        name.setText("Hey, " + myDb.getName());


        SharedPreferences timerPrefs = getActivity().getSharedPreferences("timerPrefs", Context.MODE_PRIVATE);
        boolean firstStart = timerPrefs.getBoolean("firstStart", true);
        if (firstStart) {

            addInitialCodes();
        }


        addCurrCodes(currStartCodes,"start_codes_table");
        addCurrCodes(currStopCodes,"stop_codes_table");
        originalStartCodes = addNewCodes("startcodes.txt");
        originalStopCodes = addNewCodes("stopcodes.txt");

        checkForUpdates(currStartCodes, originalStartCodes, "start_codes_table");
        checkForUpdates(currStopCodes,originalStopCodes, "stop_codes_table");

            startButton.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(final View v) {
                    if(editText.length() == 0){
                    toastMessage("No code was entered, please try again");
                    }
                    else if (currStartCodes.contains(editText.getText().toString())) {
                        currStartCodes.remove(editText.getText().toString());
                        myDb.deleteCode("start_codes_table", editText.getText().toString());
                        stopwatch.setBase(SystemClock.elapsedRealtime());
                        stopwatch.start();
                        startButton.setVisibility(v.INVISIBLE);
                        stopButton.setVisibility(v.VISIBLE);
                        stopwatch.setOnChronometerTickListener(new Chronometer.OnChronometerTickListener() {
                            @Override
                            public void onChronometerTick(Chronometer chronometer) {
                                if ((SystemClock.elapsedRealtime() - stopwatch.getBase()) >= 500000) {
                                    stopwatch.setBase(SystemClock.elapsedRealtime());
                                    stopwatch.stop();
                                    stopButton.setVisibility(v.INVISIBLE);
                                    startButton.setVisibility(v.VISIBLE);
                                    showPopup("No stop code was entered for 24 hours. The timer has been reset");
                                }
                                stopButton.setOnClickListener(new View.OnClickListener() {
                                    @Override
                                    public void onClick(View v) {
                                        if(editText.length() == 0){
                                            toastMessage("No code was entered");
                                        }
                                        if (currStopCodes.contains(editText.getText().toString())) {
                                            currStopCodes.remove(editText.getText().toString());
                                            myDb.deleteCode("stop_codes_table", editText.getText().toString());
                                            long totalTime = SystemClock.elapsedRealtime() - stopwatch.getBase();
                                            setPointsFromTime(totalTime);
                                            stopwatch.setBase(SystemClock.elapsedRealtime());
                                            stopwatch.stop();
                                            listener.onPointsTimerSent(myDb.getPoints());
                                            stopButton.setVisibility(v.INVISIBLE);
                                            startButton.setVisibility(v.VISIBLE);

                                        }
                                        else{
                                            toastMessage("The code you entered is not valid, please try again");
                                        }
                                    }
                                });
                            }
                        });
                    }
                    else{
                        toastMessage("The code you entered is not valid, please try again");
                    }
                }
            });


        return v;

    }


    public void updatePoints(int newPoints){
        points.setText(String.valueOf(newPoints));
    }

    public void checkForUpdates(List<String> currCodes, List<String> originalCodes, String table){
        List<String> tempCodes = new ArrayList<>();
        for(int i = 0; i<currCodes.size(); i++){
            for (int j = 0; j<originalCodes.size(); j++){
                if(originalCodes.get(j).equals(currCodes.get(i))){
                    tempCodes.add(currCodes.get(i));
                }
            }
        }
        if(!(currCodes.equals(tempCodes))){
            myDb.updateCodes(table,originalCodes);
        }

    }
    private void addCurrCodes(List<String> codeList, String table) {
        Cursor c = myDb.getAllData("codes", table);
        c.moveToFirst();
        while(!c.isAfterLast()) {
            codeList.add(c.getString(c.getColumnIndex("codes")));
            c.moveToNext();
        }
    }

    public void setPointsFromTime(long totalTime){
        int pointsEarned = 0;
        int minutes = (int) ((totalTime/1000) /60);
        if(totalTime > 10000 && totalTime < 20000){
            pointsEarned = 10;
            myDb.addPoints(pointsEarned);
            points.setText(String.valueOf(myDb.getPoints()));

        }
        else if(totalTime >= 20000 && totalTime < 40000){
            pointsEarned = 50;
            myDb.addPoints(pointsEarned);
            points.setText(String.valueOf(myDb.getPoints()));
        }
        else if(totalTime >= 40000 && totalTime < 60000){
            pointsEarned = 75;
            myDb.addPoints(pointsEarned);
            points.setText(String.valueOf(myDb.getPoints()));
        }
        else if(totalTime >= 60000 && totalTime < 80000){
            pointsEarned = 125;
            myDb.addPoints(pointsEarned);
            points.setText(String.valueOf(myDb.getPoints()));
        }
        else if(totalTime >= 80000 && totalTime < 100000){
            pointsEarned = 225;
            myDb.addPoints(pointsEarned);
            points.setText(String.valueOf(myDb.getPoints()));
        }
        else if(totalTime >= 100000 && totalTime < 120000){
            pointsEarned = 400;
            myDb.addPoints(pointsEarned);
            points.setText(String.valueOf(myDb.getPoints()));
        }
        else if(totalTime >= 120000 && totalTime < 140000){
            pointsEarned = 700;
            myDb.addPoints(pointsEarned);
            points.setText(String.valueOf(myDb.getPoints()));
        }
        if(minutes == 1){
            showPopup("Well done, you spent "+ minutes +" minute at the library and have earned " + pointsEarned + " points!\nYour new points balance is: " + myDb.getPoints());

        }
        else{
            showPopup("Well done, you spent "+ minutes +" minutes at the library and have earned " + pointsEarned + " points!\nYour new points balance is: " + myDb.getPoints());

        }
    }

    public void initialSetName(){
        name.setText("Hey, "+ myDb.getName());
    }

    public void showPopup(String text){
        popup = new Dialog(getActivity());
        popup.getWindow().setBackgroundDrawable(new ColorDrawable(Color.TRANSPARENT));
        popup.setContentView(R.layout.popup_layout);
        ImageView closeBtn = popup.findViewById(R.id.closeBtn);
        TextView popupText = popup.findViewById(R.id.popupText);
        setTextToEdit(text);
        popupText.setText(getTextToEdit());
        closeBtn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                popup.dismiss();
            }
        });
        popup.show();

    }
    private List<String> addNewCodes(String path){
        List<String> newList;
        listFromFile = new ListFromFile(getActivity().getApplicationContext());
        newList = listFromFile.readLine(path);
        for (String s : newList)
            Log.d(TAG, s);
        return newList;
    }

    private void addInitialCodes(){
        List<String> startList;
        listFromFile = new ListFromFile(getActivity().getApplicationContext());
        startList = listFromFile.readLine("startcodes.txt");
        for (String s : startList)
            Log.d(TAG, s);

        myDb.storeCodes(startList, "start_codes_table");

        List<String> stopList;
        stopList = listFromFile.readLine("stopcodes.txt");
        for (String d : stopList)
            Log.d(TAG, d);

        myDb.storeCodes(stopList, "stop_codes_table");

        SharedPreferences timerPrefs = getActivity().getSharedPreferences("timerPrefs", Context.MODE_PRIVATE);
        SharedPreferences.Editor editor = timerPrefs.edit();
        editor.putBoolean("firstStart", false);
        editor.apply();
    }

    public void setTextToEdit(String textToEdit) {
        this.textToEdit = textToEdit;
    }

    public String getTextToEdit() {
        return textToEdit;
    }


    public void toastMessage(String message){
        Toast.makeText(getActivity().getApplicationContext(),message,Toast.LENGTH_LONG).show();
    }

    @Override
    public void onAttach(@NonNull Context context) {
        super.onAttach(context);
        if (context instanceof TimerListener) {
            listener = (TimerListener) context;
        }
        else{
            throw new RuntimeException(context.toString() + "must implement TimerListener");
        }
    }

    @Override
    public void onDetach() {
        super.onDetach();
        listener = null;
    }
}

File: app/src/main/java/com/example/librewards/TimerFragment.kt
/*Author: Arman Jalilian
Date of Completion: 07/06/2020
Module Code: CSC3122
Application Name: Lib Rewards
Application Purpose: Rewards students as they spend time at the library
Class Name: TimerFragment
Class Purpose: The starting fragment for when a user opens the application. It allows the user to enter start and stop code given by
library staff for the day. The start code will start the timer and the stop code will stop. The duration spent at the library is then
converted into points where a user can later redeem in RewardsFragment.
 */
package com.example.librewards

import android.app.Dialog
import android.content.Context
import android.graphics.Color
import android.graphics.drawable.ColorDrawable
import android.os.Bundle
import android.os.SystemClock
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.*
import android.widget.Chronometer.OnChronometerTickListener
import androidx.fragment.app.Fragment
import com.google.firebase.database.*
import java.util.*

class TimerFragment : Fragment() {
    var popup: Dialog? = null
    var stopwatch: Chronometer? = null
    var myDb: DatabaseHelper? = null
    private var editText: EditText? = null
    var textToEdit: String? = null
    private var totalTime: Long? = null
    private var startButton: Button? = null
    private var stopButton: Button? = null
    private var points: TextView? = null
    private var name: TextView? = null
    var listener: TimerListener? = null
    lateinit var database: DatabaseReference
    lateinit var fh: FirebaseHandler
    private lateinit var mainActivity: MainActivity
    private lateinit var adminActivity: AdminActivity

    //Interface that consists of a method that will update the points in "RewardsFragment"
    interface TimerListener {
        fun onPointsTimerSent(points: Int)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        // Inflate the layout for this fragment
        val v = inflater.inflate(R.layout.fragment_timer, container, false)
        //Assigns the field to the view's specified in the fragment_timer XML file file
        mainActivity = activity as MainActivity
        adminActivity = AdminActivity()

        stopwatch = v.findViewById(R.id.stopwatch)
        editText = v.findViewById(R.id.startText)
        startButton = v.findViewById(R.id.startButton)
        stopButton = v.findViewById(R.id.stopButton)
        myDb = DatabaseHelper(requireActivity().applicationContext)
        points = v.findViewById(R.id.points)
        points?.text = myDb?.myPoints.toString()
        name = v.findViewById(R.id.nameTimer)
        fh = FirebaseHandler()
        database = FirebaseDatabase.getInstance().reference
        //Sets the name of user for this fragment by retrieving it from the database
        val wholeName = getString(R.string.Hey) + " " + myDb?.name
        name?.text = wholeName

        //Creating a preference for activity on first start-up only

        //Anything enclosed in the 'if' statement will only run once; at first start-up.
        if (firstStart()) {
        }

        return v
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        addTimerEventListener(database)

    }

    private fun addTimerEventListener(timerReference: DatabaseReference) {
        val refChild = fh.getChild("users",mainActivity.email, "studying")
        var isStudying: String
        val timerListener = object : ValueEventListener {
            override fun onDataChange(dataSnapshot: DataSnapshot) {
                isStudying = dataSnapshot.value.toString()
                Log.d("TAG", isStudying)
                if (isStudying == "1") {
                    stopwatch?.base = SystemClock.elapsedRealtime()
                    stopwatch?.start()
                    stopwatch?.onChronometerTickListener = OnChronometerTickListener {
                        //Checks if the stopwatch has gone over 24 hours. If so, the stopwatch resets back to its original state
                        if (SystemClock.elapsedRealtime() - stopwatch!!.base >= 800000) {
                            stopwatch?.base = SystemClock.elapsedRealtime()
                            stopwatch?.stop()
                            stopButton?.visibility = View.INVISIBLE
                            startButton?.visibility = View.VISIBLE
                            showPopup("No stop code was entered for 24 hours. The timer has been reset")
                        }

                    }

                } else if (isStudying == "0") {
                    totalTime = SystemClock.elapsedRealtime() - stopwatch!!.base
                    setPointsFromTime(totalTime!!)
                    stopwatch?.base = SystemClock.elapsedRealtime()
                    stopwatch?.stop()
                    //Listener to communicate with Rewards Fragment and give the points to display in there
                    listener?.onPointsTimerSent(myDb!!.myPoints)
                    refChild.setValue("2")
                }

            }

            override fun onCancelled(error: DatabaseError) {
                // Failed to read value
                Log.w("TAG", "Failed to read value.", error.toException())
            }
        }
        refChild.addValueEventListener(timerListener)
    }

    //Method that converts the duration spent at the library into points
    private fun setPointsFromTime(totalTime: Long) {
        val minutes = (totalTime / 1000 / 60).toInt()
        val pointsEarned: Int = when (totalTime) {
            in 0..10000 -> 0
            in 10001..29999 -> 10
            in 30000..59999 -> 50
            in 60000..119999 -> 75
            in 120000..179999 -> 125
            in 180000..259999 -> 225
            in 260000..399999 -> 400
            else -> 700
        }
        myDb?.addPoints(pointsEarned)
        points?.text = myDb?.myPoints.toString()
        if (minutes == 1) {
            showPopup("Well done, you spent $minutes minute at the library and have earned $pointsEarned points! Your new points balance is: ${myDb?.myPoints}")
        } else {
            showPopup("Well done, you spent $minutes minutes at the library and have earned $pointsEarned points! Your new points balance is:  ${myDb?.myPoints}")
        }
    }

    private fun firstStart(): Boolean {
        val timerPrefs = requireActivity().getSharedPreferences("timerPrefs", Context.MODE_PRIVATE)
        return timerPrefs.getBoolean("firstStart", true)
    }


    //Method to set the name on first start-up. Method is called in MainActivity
    fun initialSetName() {
        name?.text = getString(R.string.Hey) + " " + myDb?.name
    }

    fun updatePoints(newPoints: Int) {
        points!!.text = newPoints.toString()
    }

    //Method that creates a popup
    private fun showPopup(text: String?) {
        popup = Dialog(requireActivity())
        popup?.window?.setBackgroundDrawable(ColorDrawable(Color.TRANSPARENT))
        popup?.setContentView(R.layout.popup_layout)
        val closeBtn = popup?.findViewById<ImageView>(R.id.closeBtn)
        val popupText = popup?.findViewById<TextView>(R.id.popupText)
        textToEdit = text
        popupText?.text = textToEdit
        closeBtn?.setOnClickListener { popup?.dismiss() }
        popup?.show()
    }

    //Custom Toast message
    fun toastMessage(message: String?) {
        Toast.makeText(requireActivity().applicationContext, message, Toast.LENGTH_LONG).show()
    }

    override fun onAttach(context: Context) {
        super.onAttach(context)
        listener = if (context is TimerListener) {
            context
        } else {
            throw RuntimeException(context.toString() + "must implement TimerListener")
        }
    }

    override fun onDetach() {
        super.onDetach()
        listener = null
    }

    companion object {
        val TAG = TimerFragment::class.java.simpleName
    }
}